nvflare/fuel/f3/sfm/conn_manager.py

```diff
@@ -0,0 +1,452 @@
+        self.frame_mgr_executor = ThreadPoolExecutor(FRAME_THREAD_POOL_SIZE, "frame_mgr")
+        self.lock = threading.Lock()
+        self.null_conn = NullConnection()
+        self.send_frame_stats = StatsPoolManager.add_time_hist_pool(
+            "sfm_send_frame", "SFM send_frame time in secs", scope=local_endpoint.name
+        )
+
+    def add_connector(self, driver: Driver, params: dict, mode: Mode) -> str:
+
+        # Validate parameters
+        capabilities = driver.capabilities()
+        support_ssl = capabilities.get(DriverCap.SUPPORT_SSL, False)
+
+        if ssl_required(params) and not support_ssl:
+            scheme = params.get(DriverParams.SCHEME.value, "Unknown")
+            raise CommError(
+                CommError.BAD_CONFIG,
+                f"Connector with scheme {scheme} requires SSL but " f"driver {driver.get_name()} doesn't support it",
+            )
+
+        handle = get_handle()
+        connector = Connector(handle, driver, params, mode, 0, 0, False, False)
+        driver.register_conn_monitor(self)
+        with self.lock:
+            self.connectors[handle] = connector
+
+        log.debug(f"Connector {connector} is created")
+
+        if self.started:
+            self.start_connector(connector)
+
+        return handle
+
+    def remove_connector(self, handle: str):
+        with self.lock:
+            connector = self.connectors.pop(handle, None)
+            if connector:
+                connector.stopping = True
+                connector.driver.shutdown()
+                log.debug(f"Connector {connector} is removed")
+            else:
+                log.error(f"Unknown connector handle: {handle}")
+
+    def start(self):
+        with self.lock:
+            for handle in sorted(self.connectors.keys()):
+                connector = self.connectors[handle]
+                if not connector.started:
+                    self.start_connector(connector)
+
+        self.started = True
+
+    def stop(self):
+
+        with self.lock:
+            for handle in sorted(self.connectors.keys()):
+                connector = self.connectors[handle]
+                connector.stopping = True
+                connector.driver.shutdown()
+
+        self.conn_mgr_executor.shutdown(True)
+        self.frame_mgr_executor.shutdown(True)
+
+    def find_endpoint(self, name: str) -> Optional[Endpoint]:
+
+        sfm_endpoint = self.endpoints.get(name)
+        if not sfm_endpoint:
+            log.debug("Endpoint {name} doesn't exist")
+            return None
+
+        return sfm_endpoint.endpoint
+
+    def get_connections(self, name: str) -> Optional[List[SfmConnection]]:
+
+        sfm_endpoint = self.endpoints.get(name)
+        if not sfm_endpoint:
+            log.debug("Endpoint {name} doesn't exist")
+            return None
+
+        return sfm_endpoint.connections
+
+    def send_message(self, endpoint: Endpoint, app_id: int, headers: Headers, payload: BytesAlike):
```

Comment:
I see we define a `Message` class up there which wraps the `headers` and `payload`.
In that case, could we change the method signature to just send `Message`?

Reply:
This is intentional. The message is only used in Communicator. Once in SFM, it only deals with dict and bytes, no Messages.